File: common/src/main/java/dev/obscuria/fragmentum/network/PayloadCodec.java
package dev.obscuria.fragmentum.network;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.mojang.serialization.Codec;
import com.mojang.serialization.JsonOps;
import io.netty.handler.codec.DecoderException;
import io.netty.handler.codec.EncoderException;
import net.minecraft.Util;
import net.minecraft.core.RegistryAccess;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.RegistryOps;
import net.minecraft.util.GsonHelper;

import java.util.function.Supplier;

@SuppressWarnings("unused")
public interface PayloadCodec<T> {

    static <T> PayloadCodec<T> registryFriendly(Codec<T> codec, Supplier<RegistryAccess> registryAccess) {
        return new RegistryFriendly<>(codec, registryAccess);
    }

    void write(FriendlyByteBuf buf, T value);

    T read(FriendlyByteBuf buf);

    record RegistryFriendly<T>(Codec<T> codec, Supplier<RegistryAccess> registryAccess) implements PayloadCodec<T> {

        private static final int MAX_LENGTH = 1048576;
        private static final Gson GSON = new Gson();

        @Override
        public void write(FriendlyByteBuf buf, T value) {
            final var ops = RegistryOps.create(JsonOps.INSTANCE, registryAccess.get());
            buf.writeUtf(GSON.toJson(Util.getOrThrow(codec.encodeStart(ops, value), exception -> new EncoderException("Failed to encode: %s %s".formatted(exception, value)))), MAX_LENGTH);
        }

        @Override
        public T read(FriendlyByteBuf buf) {
            final var element = GsonHelper.fromJson(GSON, buf.readUtf(MAX_LENGTH), JsonElement.class);
            final var ops = RegistryOps.create(JsonOps.INSTANCE, registryAccess.get());
            return Util.getOrThrow(codec.parse(ops, element), exception -> new DecoderException("Failed to decode json: %s".formatted(exception)));
        }
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/network/PayloadCodec.kt
@file:Suppress("unused", "RedundantNullableReturnType")

package dev.obscuria.fragmentum.network

import com.google.gson.Gson
import com.google.gson.JsonElement
import com.mojang.serialization.Codec
import com.mojang.serialization.JsonOps
import io.netty.handler.codec.DecoderException
import io.netty.handler.codec.EncoderException
import net.minecraft.Util
import net.minecraft.core.RegistryAccess
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.resources.RegistryOps
import net.minecraft.util.GsonHelper

sealed interface PayloadCodec<T> {

    fun write(buf: FriendlyByteBuf, value: T & Any)

    fun read(buf: FriendlyByteBuf): T

    private data class RegistryFriendly<T>(
        val codec: Codec<T>,
        val registry: () -> RegistryAccess
    ) : PayloadCodec<T> {

        override fun write(buf: FriendlyByteBuf, value: T & Any) {
            buf.writeRegistryFriendly(codec, registry(), value)
        }

        override fun read(buf: FriendlyByteBuf): T {
            return buf.readRegistryFriendly(codec, registry())
        }
    }

    companion object {

        private const val MAX_LENGTH = 1048576
        private val GSON: Gson = Gson()

        fun <T> registryFriendly(codec: Codec<T>, registry: () -> RegistryAccess): PayloadCodec<T> {
            return RegistryFriendly(codec, registry)
        }

        fun <T> FriendlyByteBuf.writeRegistryFriendly(codec: Codec<T>, registry: RegistryAccess, value: T) {
            val registryOps = RegistryOps.create(JsonOps.INSTANCE, registry)
            codec.encodeStart(registryOps, value).let {
                this.writeUtf(GSON.toJson(Util.getOrThrow(it) { exception ->
                    EncoderException("Failed to encode: $exception ${value.toString()}")
                }), MAX_LENGTH)
            }
        }

        fun <T> FriendlyByteBuf.readRegistryFriendly(codec: Codec<T>, registry: RegistryAccess): T {
            val element = GsonHelper.fromJson(GSON, this.readUtf(MAX_LENGTH), JsonElement::class.java)
            val registryOps = RegistryOps.create(JsonOps.INSTANCE, registry)
            return codec.parse(registryOps, element).let {
                Util.getOrThrow(it) { exception ->
                    DecoderException("Failed to decode json: $exception")
                }
            }
        }
    }
}